### PR TITLE
Bring Dotnet AutoGen

### PR DESCRIPTION
**Pull request by LittleLittleCloud**
## Related issue number

#48 

## Checks

- [ ] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

This PR adds AutoGen dotnet implementation to the repo, the implementation contains three libraries
- `AutoGen` The Dotnet AutoGen library
- `AutoGen.SourceGenerator` The type-safe function call source generator
- `AutoGen.DotnetInteractive` Add support for running dotnet code.

### (In preview) website for API Doc

- https://microsoft.github.io/autogen-for-net/

All code lives under /dotnet folder, project structure is as follows
- eng: infrustructure && config
- src: where library lives
- sample: dotnet AutoGen samples
- test: test code

Please kindly review the provided samples and offer feedback on the API implementation. Given the differences in language and ecosystem, achieving a syntax-level match on dotnet APIs is highly unlikely, and fulfilling the requirement of 100% functionality match may also be challenging. Nevertheless, our primary objective is to achieve a semantic-level match to the greatest extent possible.

### How the package will be distributed?
We plan to publish both nightly-build and official releases. The official release will be published to nuget.org, and nightly-build will be published to a third-party feed once this PR get merged.

### CODE REVIEW

**Review by erhan0 — APPROVED** [review]
This looks solid. Any feedback I have is stylistic.

**Review by ekzhu — COMMENTED** [review]
This is a one big PR so I have some initial thoughts after skimming through the code:

1. The class names and package names do not match the current Python classes directly. This may lead to future maintenance issues if we want to sync functionalities. I think the Python core will go through several design iterations, so this may also lead to more instability in the dotnet API.   
2. The feature parity is a bit too large. e.g., ConfigListFromJson is not implemented. Is there an urgent use case that requires the dotnet version?

**Comment by LittleLittleCloud** [comment]
@ekzhu 

> The class names and package names do not match the current Python classes directly.

Yeah, I was also puzzed by this question while working on this PR. On one hand, as I addressed in the PR description, Implementing a syntax-matching dotnet AutoGen poses challenges due to the notable differences between Python and CSharp. For instance, Python AutoGen widely uses the `Union` type in constructs like AssistantAgent, accepting parameters with different types, which is not available in CSharp. Additionally, decorators are exclusive to Python.

On the other hand,  striving for a syntax-matching dotnet AutoGen might also not be the most suitable approach for dotnet developers. Dotnet has a very different code style compared to Python. For example, in naming style, dotnet uses camel case and Python prefers snake case. And in the async API style, dotnet's prefers the `Async` suffix and python prefers `a_` prefix. Moreover, dotnet employs a static type system, while Python relies on duck typing. Attempting a syntax match could lead to the awkward use of dynamic or object types in the code, conflicting with the typical coding style familiar to dotnet developers.

To address this dilemma, my two cents here is to focus on achieving a semantic-match on API level between python and dotnet AutoGen, rather than strictly adhering to syntax. The distinction lies in the fact that semantic matching accommodates differences arising from coding styles, whereas syntax matching does not. Additionally, for differences caused by language features (such as `Union` or decorators in python, and `attribution`, source generator in csharp), semantic matching allows each language to make the most of its own unique features.

For example, here's a few python vs dotnet AutoGen snippets, in semantic-match style:

<details>
<summary>two agent chat </summary>

```python
# python
import autogen

config_list = autogen.config_list_from_json(
    "OAI_CONFIG_LIST",
    filter_dict={
        "model": ["gpt-4", "gpt4", "gpt-4-32k", "gpt-4-32k-0314", "gpt-4-32k-v0314"],
    },
)

# Create the agent that uses the LLM.
assistant = AssistantAgent("agent", llm_config={"config_list": config_list})

# Create the agent that represents the user in the conversation.
user_proxy = UserProxyAgent("user", code_execution_config=False)

assistant.a_send(message="How can I help you today?", recipient: user_proxy)
```

```csharp
// dotnet
using AutoGen;
var filterConfig = new FilterConfig
{
   Model=new [] {"gpt-4", ....},
};
var configList = AutoGen.ConfigListFromJson("xxx", filterConfig: filterConfig );
var assistantAgentConfig = new AssistantAgentConfig
{
   ConfigList = configList
};
var assistant = new AssistantAgent("agent",llmConfig: assistantAgentConfig);
var userProxy = new UserProxyAgent("user");
var history = await assistant.SendAsync(message: "How can I help you today?", receiver: userProxy);
```
</details>


<details>
<summary>two agent chat function call</summary>

```python
llm_config = {
    "config_list": config_list,
    "timeout": 120,
}

chatbot = autogen.AssistantAgent(
    name="chatbot",
    system_message="For currency exchange tasks, only use the functions you have been provided with. Reply TERMINATE when the task is done.",
    llm_config=llm_config,
)

# create a UserProxyAgent instance named "user_proxy"
user_proxy = autogen.UserProxyAgent(
    name="user_proxy",
    is_termination_msg=lambda x: x.get("content", "") and x.get("content", "").rstrip().endswith("TERMINATE"),
    human_input_mode="NEVER",
    max_consecutive_auto_reply=10,
)

@user_proxy.register_for_execution()
@chatbot.register_for_llm(description="Currency exchange calculator.")
def currency_calculator(
    base_amount: Annotated[float, "Amount of currency in base_currency"],
    base_currency: Annotated[CurrencySymbol, "Base currency"] = "USD",
    quote_currency: Annotated[CurrencySymbol, "Quote currency"] = "EUR",
) -> str:
    quote_amount = exchange_rate(base_currency, quote_currency) * base_amount
    return f"{quote_amount} {quote_currency}"
```

```csharp
// in dotnet, we can rely on structured comments and source generator to generate function_call description.
// however, it's challenging to achieve a syntax-match with python version in this case, because both decorator and Pydantic module are not available in dotnet.

// source generator will generate CurrencyCalculatorFunctionDefinition and CurrencyCalculatorWrapper for further consuming
public partial class Function
{
   /// <summary>
   /// Currency exchange calculator.
   /// </summary>
   /// <param name="baseAmount">Amount of currency in base_currency</param>
   /// <param name="baseCurrency">Base currency</param>
   /// <param name="quoteCurrency">Quote currency</param>
   [FunctionAttribution]
   public async Task<string> CurrencyCalculator(float baseAmount, string baseCurrency, string quoteCurrency)
   {
       // implementation...
   }
}
var functions = new Function();
var assistantAgentConfig = new AssistantAgentConfig
{
   ConfigList = configList,
   FunctionDefinitions = new [] {
        functions.CurrencyCalculatorFunctionDefinition,
   }
};

var chatbot = new AssistantAgent(
    name: "chatbot",
    systemMessage: "For currency exchange tasks, only use the functions you have been provided with. Reply TERMINATE when the task is done.",
    llmConfig: assistantAgentConfig );

# create a UserProxyAgent instance named "user_proxy"
var userProxy = new UserProxyAgent(
    name: "user_proxy",
    isTermination: async (msg, ct) => msg.EndsWith("Termination")
    humanInputMode: humanInputMode.Never,
    functionMap: new Dictionary<string, Func<string, Task<string>>>
            {
                { nameof(Function.CurrencyCalculator), function.CurrencyCalculatorWrapper },
            });
)
```

> The feature gap is a bit too large. e.g., ConfigListFromJson is not implemented -- of course it's not absolutely required, but right now this is not clear.

That's true, In this PR, the features to be implemented are mostly limited to `agentchat`, for example
- `AssistantAgent`
- (wip) `UserProxyAgent`
- dynamic group chat

So most APIs under `oai` are still missing. For example
- config_list_from_models
- config_list_from_json
- ...

You can reach out to [README.md](./dotnet/README.md) for a complete feature list included in this PR, These features are selected based on some of my previous work (like AgentChat or MLNet-task-101) so it will be very subjective. Feel free to add more features to that list if you think it's necessary to port together in this PR as well.

> Perhaps we can start with a concrete use case for dotnet and implement the relevant interfaces that match the current Python version?

I put together all use cases under `dotnet/sample/AutoGen.BasicSamples`, maybe we can start shaping out API there?

**Comment by ekzhu** [comment]
You are right. It is not possible to have exact syntax match between Python and dotnet. I was thinking in terms of semantic similarity. 

I think perhaps we can document some example usage cases in the README.md file -- like what you did in the previous comment. That makes it much easier for people to get started. 

Is there a reason to have a dedicated class for assistant agent config? Maybe simply call it `LLMConfig` to match the Python version? 

For UserProxyAgent, I assume we are still generating Python code right?

**Comment by ekzhu** [comment]
The items in the README.md looks good to me. If they are not completed, then perhaps it is better to set the PR status to Draft.

**Comment by LittleLittleCloud** [comment]
> I think perhaps we can document some example usage cases in the README.md file -- like what you did in the previous comment. That makes it much easier for people to get started.

That's a good call. I'll update readme to include usage cases

> Is there a reason to have a dedicated class for assistant agent config? Maybe simply call it LLMConfig to match the Python version?

Having a dedicated class for assistant agent config allows us to avoid using `object` or `IDictionary<string, object>` as parameter type when passing `llm_config`, which I feel like to be a more natural way when writing dotnet?

The reason not using `LLMConfig` is because there's already an `ILLMConfig` for `llm_config.config_list` field.

> For UserProxyAgent, I assume we are still generating Python code right?

Are you talking about running Python code by UserProxyAgent? As there's no limitation on which programming language to generate by an agent since it's all about prompting.

For dotnet AutoGen, we have the option to run dotnet code. The functionality is provided by `AutoGen.DotnetInteractive` package and the usage example is [dotnet/sample/AutoGen.BasicSamples/Example04_Dynamic_GroupChat_Get_MLNET_PR.cs](https://github.com/microsoft/autogen/pull/924/files#diff-d2fe4788de7878793548f4c4b18d7ef56f640141104654abb3aedd019e07af93). 

Running python code is not currently supported in dotnet AutoGen because it needs Python interpreter to be installed. If we want to support that feature, it is probably better for us to only support the `docker` mode and distribute the function inside another package, like `AutoGen.PythonInterpreter`

**Comment by ekzhu** [comment]
> Having a dedicated class for assistant agent config allows us to avoid using object or IDictionary<string, object> as parameter type when passing llm_config, which I feel like to be a more natural way when writing dotnet? The reason not using LLMConfig is because there's already an ILLMConfig for llm_config.config_list field.

Makes sense. 


> For dotnet AutoGen, we have the option to run dotnet code. The functionality is provided by AutoGen.DotnetInteractive package and the usage example is [dotnet/sample/AutoGen.BasicSamples/Example04_Dynamic_GroupChat_Get_MLNET_PR.cs](https://github.com/microsoft/autogen/pull/924/files#diff-d2fe4788de7878793548f4c4b18d7ef56f640141104654abb3aedd019e07af93).
I was trying to run this and got: 

```
Unhandled exception. System.NullReferenceException: Object reference not set to an instance of an object.
   at InteractiveService.RestoreDotnetInteractive() in /Users/ekzhu/autogen/dotnet/src/AutoGen.DotnetInteractive/InteractiveService.cs:line 182
   at InteractiveService.CreateKernelAsync(String workingDirectory, CancellationToken ct) in /Users/ekzhu/autogen/dotnet/src/AutoGen.DotnetInteractive/InteractiveService.cs:line 138
   at InteractiveService.StartAsync(String workingDirectory, CancellationToken ct) in /Users/ekzhu/autogen/dotnet/src/AutoGen.DotnetInteractive/InteractiveService.cs:line 38
   at Example04_Dynamic_GroupChat_Get_MLNET_PR.RunAsync() in /Users/ekzhu/autogen/dotnet/sample/AutoGen.BasicSamples/Example04_Dynamic_GroupChat_Get_MLNET_PR.cs:line 70
   at Program.<Main>$(String[] args) in /Users/ekzhu/autogen/dotnet/sample/AutoGen.BasicSamples/Program.cs:line 4
   at Program.<Main>(String[] args)
zsh: abort      sample/AutoGen.BasicSamples/bin/Debug/net7.0/AutoGen.BasicSample
```

**Comment by LittleLittleCloud** [comment]
@ekzhu I just push a fix, could you try again

**Comment by bitnom** [comment]
This is an amazing effort. I have concerns about merging a separate language port into the repo. Having only just been made aware of the issue thread and PR, my concern at present is basically 'We should have concerns.'

**Comment by LittleLittleCloud** [comment]
> This is an amazing effort. I have concerns about merging a separate language port into the repo. Having only just been made aware of the issue thread and PR, my concern at present is basically 'We should have concerns.'

Thank you for addressing this concern. We can draw on  [semantic-kernel](https://github.com/microsoft/semantic-kernel)'s experience to manage issues/PRs and distributing packages for different language SDKs within a monorepo.

For instance, we can use the following approach(es) to address tagging/labeling issues for different langauge.
- Labeling PRs: [label-pr.yml](https://github.com/microsoft/semantic-kernel/blob/main/.github/workflows/label-prs.yml)
- Labeling issues: [label-issues.yml](https://github.com/microsoft/semantic-kernel/blob/main/.github/workflows/label-issues.yml)

File: website/blog/2023-10-18-RetrieveChat/img/autogen-rag.gif
Comment (sonichi) [inline website/blog/2023-10-18-RetrieveChat/img/autogen-rag.gif:1]: Is this change needed?

**Comment by sonichi** [comment]
Is this test required to run: https://github.com/microsoft/autogen/actions/runs/8839202092/job/24272062128?pr=924 ?

**Comment by LittleLittleCloud** [comment]
> Is this test required to run: https://github.com/microsoft/autogen/actions/runs/8839202092/job/24272062128?pr=924 ?

No, it's optional.

File: website/blog/2023-10-18-RetrieveChat/img/autogen-rag.gif
Comment (LittleLittleCloud) [reply website/blog/2023-10-18-RetrieveChat/img/autogen-rag.gif:1]: actually it should be needed because .gif also needs to be saved to lfs. But not in this PR.

I'll revert this change

**Review by ekzhu — APPROVED** [review]
